### ISSUE TITLE
Compiler versions are very confused by translation to opam next

### ISSUE THREAD

**Issue by Drup** [ocaml/opam#2539]
opam doesn't now the compiler versions of my old switches anymore:
```
# switch                      compiler      description
3.12.1                                      
4.01.0                                      
4.02.0+modular-implicits                    
4.02.1+BER                                  
4.02.1+modular-implicits-ber                
4.02.3                                      
4.03.0                        ocaml.4.03.0  Official 4.03.0 release
4.03.0+local-git-trunk                      
eliom                         ocaml.4.02.3  Official 4.02.3 release
last                                        
latest                        ocaml.4.02.3  Official 4.02.3 release
mirage                                      
system                                      
termite                                     
```

**Comment by AltGr** [comment]
Could you check the contents of `~/.opam/<switch>/.opam-switch/switch-state` ? You should have a field looking like:
```
compiler: [
  "base-bigarray.base"
  "base-ocamlbuild.base"
  "base-threads.base"
  "base-unix.base"
  "ocaml.4.02.3"
]
```
The `opam switch` command looks in this list for a package having the `compiler` flag set, so it could also be the case that the translated definition `~/.opam/<switch>/.opam-switch/packages/ocaml.<version>/opam` doesn't properly have `flags: compiler`.

Thanks for the feedback !

**Comment by Drup** [comment]
```
% cat .opam/4.02.3/.opam-switch/switch-state 
opam-version: "2.0"
compiler: [
  "base-bigarray.~unknown"
  "base-ocamlbuild.~unknown"
  "base-threads.~unknown"
  "base-unix.~unknown"
  "ocaml.4.02.3"
]
roots: ["ocaml.4.02.3"]
installed: ["ocaml.4.02.3"]
```

`~/.opam/4.02.3/.opam-switch/packages/ocaml` doesn't exist.

**Comment by Drup** [comment]
Actually, `~/.opam/4.02.3/.opam-switch/packages` is empty!

**Comment by AltGr** [comment]
Arg, so that's why then... now since your upgrade didn't go smoothly, this may be why, but I'll need to make sure that it always goes smoothly at least from 1.2.2
Do you still have a `~/.opam/packages` or `~/.opam/compilers` ? (they should have been removed)

**Comment by Drup** [comment]
I don't have a `~/.opam/packages` but I do have a `~/.opam/compilers` full of things. Should I remove it ?

**Comment by AltGr** [comment]
You can, it won't be used anymore; I'll check that the upgrade properly removes it.

**Comment by AltGr** [comment]
(it didn't: fixed, thanks; of course that won't help you :p)

Migration of installed compilers should have been handled by [this function](https://github.com/ocaml/opam/blob/next/src/state/opamGlobalState.ml#L200)

**Comment by AltGr** [comment]
Fixed by #2554
